File: ts-log/src/intersect.ts
/**
 * The pairwise intersection a CAS loser runs (15): subsumed, disjoint in
 * the strict sense (zero shared keys of any class, commute cells
 * included), a quantitative W test on parent keys shared child-to-child,
 * and conflict for everything else. The winner's footprint is always a
 * recomputation from its fetched ops — the carried-and-checked law —
 * so this function takes ops, never published sections.
 */

import type { LogTheory } from "#descriptor.ts"
import { descriptorOf } from "#descriptor.ts"
import type { BatchOp, FootprintRich } from "#footprint.ts"
import { computeFootprint, keyIdentity } from "#footprint.ts"

interface SharedKey {
	readonly class: "F" | "K" | "C" | "W"
	readonly statement: number | undefined
	readonly keyHex: string
}

interface DeltaInterval {
	readonly lo: bigint
	readonly hi: bigint
}

interface SharedCapacityParent {
	readonly statement: number
	readonly keyHex: string
	readonly loser: DeltaInterval
	readonly winner: DeltaInterval
}

type Intersection =
	| { readonly tag: "subsumed" }
	| { readonly tag: "disjoint" }
	| { readonly tag: "capacity"; readonly parents: readonly SharedCapacityParent[] }
	| { readonly tag: "conflict"; readonly shared: readonly SharedKey[] }

function intersectRich(loser: FootprintRich, winner: FootprintRich): Intersection {
	const winnerByIdentity = new Map<string, (typeof winner.entries)[number]>()
	const winnerKeys = new Map<string, (typeof winner.entries)[number][]>()
	for (const entry of winner.entries) {
		winnerByIdentity.set(`${keyIdentity(entry)}:${"mode" in entry ? entry.mode : ""}`, entry)
		const key = keyIdentity(entry)
		const bucket = winnerKeys.get(key)
		if (bucket === undefined) {
			winnerKeys.set(key, [entry])
		} else {
			bucket.push(entry)
		}
	}

	const loserFacts = loser.entries.filter(function facts(entry) {
		return entry.class === "F"
	})
	const subsumed =
		loserFacts.length > 0 &&
		loserFacts.every(function coveredSameMode(entry) {
			return winnerByIdentity.has(`${keyIdentity(entry)}:${"mode" in entry ? entry.mode : ""}`)
		})
	if (subsumed) {
		return { tag: "subsumed" }
	}

	const shared: SharedKey[] = []
	const capacityParents: SharedCapacityParent[] = []
	let hardConflict = false
	const seen = new Set<string>()
	for (const entry of loser.entries) {
		const key = keyIdentity(entry)
		if (seen.has(key)) {
			continue
		}
		seen.add(key)
		const winnerBucket = winnerKeys.get(key)
		if (winnerBucket === undefined) {
			continue
		}
		if (entry.class !== "W") {
			hardConflict = true
			shared.push({
				class: entry.class,
				statement: entry.class === "F" ? undefined : entry.statement,
				keyHex: key.split(":")[2] ?? ""
			})
			continue
		}
		const loserModes = loser.entries.filter(function sameKey(candidate) {
			return keyIdentity(candidate) === key
		})
		const anyParent =
			loserModes.some(function parentArm(candidate) {
				return "mode" in candidate && (candidate.mode === "parent+" || candidate.mode === "parent-")
			}) ||
			winnerBucket.some(function parentArm(candidate) {
				return "mode" in candidate && (candidate.mode === "parent+" || candidate.mode === "parent-")
			})
		if (anyParent) {
			hardConflict = true
			shared.push({ class: "W", statement: entry.statement, keyHex: key.split(":")[2] ?? "" })
			continue
		}
		const identity = `W:${entry.statement}:${key.split(":")[2] ?? ""}`
		const loserInterval = loser.intervals.get(identity)
		const winnerInterval = winner.intervals.get(identity)
		if (loserInterval === undefined || winnerInterval === undefined) {
			hardConflict = true
			shared.push({ class: "W", statement: entry.statement, keyHex: key.split(":")[2] ?? "" })
			continue
		}
		capacityParents.push({
			statement: entry.statement,
			keyHex: loserInterval.keyHex,
			loser: { lo: loserInterval.lo, hi: loserInterval.hi },
			winner: { lo: winnerInterval.lo, hi: winnerInterval.hi }
		})
	}

	if (hardConflict) {
		return { tag: "conflict", shared }
	}
	if (capacityParents.length > 0) {
		return { tag: "capacity", parents: capacityParents }
	}
	return { tag: "disjoint" }
}

/**
 * Intersects a loser's batch against a winner's, both as raw ops built
 * on the same base slot. Footprints are recomputed here, never trusted.
 */
function intersectionOf(theory: LogTheory, loserOps: readonly BatchOp[], winnerOps: readonly BatchOp[]): Intersection {
	const descriptor = descriptorOf(theory)
	return intersectRich(computeFootprint(descriptor, loserOps), computeFootprint(descriptor, winnerOps))
}

interface CapacitySlack {
	/** `ceiling − measure(base)`; null where unbounded. */
	readonly plus: bigint | null
	/** `measure(base) − floor`; null where unbounded. */
	readonly minus: bigint | null
}

/**
 * The W interval test at known slack: concurrent batches commute iff the
 * worst-case endpoints respect both bounds — `Σ max-endpoints ≤ slack⁺`
 * and `Σ min-endpoints ≥ −slack⁻` at every shared parent.
 */
function capacityCommutes(
	parents: readonly SharedCapacityParent[],
	slackOf: (statement: number, keyHex: string) => CapacitySlack
): boolean {
	return parents.every(function bounded(parent) {
		const slack = slackOf(parent.statement, parent.keyHex)
		const maxSum = parent.loser.hi + parent.winner.hi
		const minSum = parent.loser.lo + parent.winner.lo
		if (slack.plus !== null && maxSum > slack.plus) {
			return false
		}
		if (slack.minus !== null && minSum < -slack.minus) {
			return false
		}
		return true
	})
}

export type { CapacitySlack, DeltaInterval, Intersection, SharedCapacityParent, SharedKey }
export { capacityCommutes, intersectionOf, intersectRich }

File: crates/bumbledb-log/src/intersect.rs
//! The pairwise loser decision: subsumed, disjoint, or conflict. The
//! winner's footprint is recomputed from its ops — never read from the
//! carried section, so an understating writer cannot steer losers onto
//! the republish path. Disjoint is strict: no shared key of any
//! existential class, commute cells included (L6's hypothesis is false
//! without excluding them), and every shared capacity parent passes
//! the evaporation-widened interval test; everything else re-judges.

use std::collections::BTreeMap;

use bumbledb::schema::StatementId;

use crate::codec::Op;
use crate::footprint::{
    CapacityKey, CapacityProfile, Entry, FootprintError, Vocabulary, capacity_profiles, footprint,
};

/// The base-state quantities the interval test prices a shared parent
/// group against. The caller supplies them (and can re-run
/// [`capacity_cell`] against winner-updated measures without touching
/// the wire).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct BaseMeasure {
    pub measure: u64,
    pub floor: u64,
    pub ceiling: Option<u64>,
}

impl BaseMeasure {
    fn slack_up(&self) -> Option<i128> {
        self.ceiling
            .map(|ceiling| i128::from(ceiling) - i128::from(self.measure))
    }

    fn slack_down(&self) -> i128 {
        i128::from(self.measure) - i128::from(self.floor)
    }
}

/// Why a pair fails to commute.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ConflictCause {
    /// A shared fact id, either mode: op effects are not independent.
    Fact { fid: [u8; 32] },

    /// A shared key-statement determinant.
    Key {
        statement: StatementId,
        key: [u8; 32],
    },

    /// A shared containment target group.
    Containment {
        statement: StatementId,
        key: [u8; 32],
    },

    /// A shared parent group whose worst-case interval endpoints
    /// violate a slack bound.
    CapacityInterval {
        statement: StatementId,
        key: [u8; 32],
    },

    /// A parent row's own move racing the other batch at the group.
    CapacityParent {
        statement: StatementId,
        key: [u8; 32],
    },

    /// A shared parent group the caller supplied no base measure for —
    /// conservatively priced as a conflict; re-judgment is always
    /// sound.
    CapacityMeasureMissing {
        statement: StatementId,
        key: [u8; 32],
    },
}

/// The loser's three outcomes at a lost slot.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LoserDecision {
    /// Every F entry of the loser appears in the winner with the same
    /// mode: the winner already performed the loser's net effect;
    /// publish nothing.
    Subsumed,

    /// Full key disjointness: republish with a re-addressed header,
    /// verdict untouched (L7 licenses it).
    Disjoint,

    /// Anything else: rebuild the base and re-judge the recorded ops.
    Conflict(ConflictCause),
}

/// One shared parent group's verdict under the interval test.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CapacityCell {
    Commutes,
    IntervalExceeded,
    ParentRace,
    MeasureMissing,
}

/// The W matrix cell for one shared parent group: parent removal
/// tolerates only an inert other side; parent addition commutes with
/// child adds alone; child intervals must jointly respect both slack
/// bounds at their worst-case endpoints.
#[must_use]
pub fn capacity_cell(
    loser: &CapacityProfile,
    winner: &CapacityProfile,
    base: Option<&BaseMeasure>,
) -> CapacityCell {
    for (mover, other) in [(loser, winner), (winner, loser)] {
        if mover.parent_remove && !other.is_inert() {
            return CapacityCell::ParentRace;
        }
        if mover.parent_add && (other.min() < 0 || other.parent_add || other.parent_remove) {
            return CapacityCell::ParentRace;
        }
    }

    if loser.min() == 0 && loser.max() == 0 && winner.min() == 0 && winner.max() == 0 {
        return CapacityCell::Commutes;
    }

    let Some(base) = base else {
        return CapacityCell::MeasureMissing;
    };
    if let Some(slack_up) = base.slack_up()
        && loser.max() + winner.max() > slack_up
    {
        return CapacityCell::IntervalExceeded;
    }
    if loser.min() + winner.min() < -base.slack_down() {
        return CapacityCell::IntervalExceeded;
    }
    CapacityCell::Commutes
}

/// The pairwise decision. Inputs are the loser's own footprint and
/// ops plus the winner's ops; the winner's footprint and both sides'
/// capacity profiles are recomputed here.
pub fn intersect(
    vocabulary: &Vocabulary,
    loser_footprint: &[Entry],
    loser_ops: &[Op],
    winner_ops: &[Op],
    base: &BTreeMap<CapacityKey, BaseMeasure>,
) -> Result<LoserDecision, FootprintError> {
    let winner_footprint = footprint(vocabulary, winner_ops)?;

    if subsumed(loser_footprint, &winner_footprint) {
        return Ok(LoserDecision::Subsumed);
    }

    if let Some(cause) = shared_existential(loser_footprint, &winner_footprint) {
        return Ok(LoserDecision::Conflict(cause));
    }

    let loser_profiles = capacity_profiles(vocabulary, loser_ops)?;
    let winner_profiles = capacity_profiles(vocabulary, winner_ops)?;
    for (key, loser_profile) in &loser_profiles {
        let Some(winner_profile) = winner_profiles.get(key) else {
            continue;
        };
        let cell = capacity_cell(loser_profile, winner_profile, base.get(key));
        match cell {
            CapacityCell::Commutes => {}
            CapacityCell::IntervalExceeded => {
                return Ok(LoserDecision::Conflict(ConflictCause::CapacityInterval {
                    statement: key.statement,
                    key: key.key,
                }));
            }
            CapacityCell::ParentRace => {
                return Ok(LoserDecision::Conflict(ConflictCause::CapacityParent {
                    statement: key.statement,
                    key: key.key,
                }));
            }
            CapacityCell::MeasureMissing => {
                return Ok(LoserDecision::Conflict(
                    ConflictCause::CapacityMeasureMissing {
                        statement: key.statement,
                        key: key.key,
                    },
                ));
            }
        }
    }

    Ok(LoserDecision::Disjoint)
}

/// Every F entry of the loser appears in the winner with the same
/// mode. Both sections are sorted, F class first, so this is one
/// two-pointer walk over the F prefixes.
fn subsumed(loser: &[Entry], winner: &[Entry]) -> bool {
    let mut winner_facts = winner
        .iter()
        .filter_map(|entry| match entry {
            Entry::Fact { fid, mode } => Some((*fid, *mode)),
            _ => None,
        })
        .peekable();

    for entry in loser {
        let Entry::Fact { fid, mode } = entry else {
            break;
        };
        loop {
            match winner_facts.peek() {
                Some((wfid, wmode)) if wfid == fid => {
                    if wmode != mode {
                        return false;
                    }
                    break;
                }
                Some((wfid, _)) if *wfid < *fid => {
                    winner_facts.next();
                }
                _ => return false,
            }
        }
    }
    true
}

/// The first shared existential-class coordinate (F, K, or C), mode
/// ignored — the strict-disjointness scan. W coordinates are priced by
/// the interval test instead.
fn shared_existential(loser: &[Entry], winner: &[Entry]) -> Option<ConflictCause> {
    let mut l = 0;
    let mut w = 0;
    while l < loser.len() && w < winner.len() {
        let lk = loser[l].share_key();
        let wk = winner[w].share_key();
        match lk.cmp(&wk) {
            std::cmp::Ordering::Less => l += 1,
            std::cmp::Ordering::Greater => w += 1,
            std::cmp::Ordering::Equal => {
                match &loser[l] {
                    Entry::Fact { fid, .. } => {
                        return Some(ConflictCause::Fact { fid: *fid });
                    }
                    Entry::Key { statement, key } => {
                        return Some(ConflictCause::Key {
                            statement: *statement,
                            key: *key,
                        });
                    }
                    Entry::Containment { statement, key, .. } => {
                        return Some(ConflictCause::Containment {
                            statement: *statement,
                            key: *key,
                        });
                    }
                    Entry::Capacity { .. } => {}
                }
                l += 1;
            }
        }
    }
    None
}
